website/project/model.py

```diff
@@ -2829,188 +2894,302 @@ def to_json(self):
+            self.state = Sanction.APPROVED
+            self._on_complete(user)
+
+    def _on_reject(self, user, token):
+        """Early termination of a Sanction"""
+        raise NotImplementedError('Sanction subclasses must implement an #_on_reject method')
+
+    def _on_complete(self, user):
+        """When a Sanction has unanimous approval"""
+        raise NotImplementedError('Sanction subclasses must implement an #_on_complete method')
+
+    def approve(self, user, token):
         """Add user to approval list if user is admin and token verifies."""
         try:
             if self.approval_state[user._id]['approval_token'] != token:
-                raise InvalidRetractionApprovalToken('Invalid retraction approval token provided.')
+                raise InvalidSanctionApprovalToken(self.APPROVAL_INVALID_TOKEN_MESSAGE.format(self.DISPLAY_NAME))
         except KeyError:
-            raise PermissionsError('User must be an admin to disapprove retraction of a registration.')
-
+            raise PermissionsError(self.APPROVAL_NOT_AUTHORIZED_MESSAGE.format(self.DISPLAY_NAME))
         self.approval_state[user._id]['has_approved'] = True
-        if all(val['has_approved'] for val in self.approval_state.values()):
-            self.state = self.RETRACTED
+        self._on_approve(user, token)
 
-            parent_registration = Node.find_one(Q('retraction', 'eq', self))
-            parent_registration.registered_from.add_log(
-                action=NodeLog.RETRACTION_APPROVED,
-                params={
-                    'node': parent_registration._id,
-                    'retraction_id': self._id,
-                },
-                auth=Auth(user),
-            )
-            # Remove any embargoes associated with the registration
-            if parent_registration.embargo_end_date or parent_registration.pending_embargo:
-                parent_registration.embargo.state = self.CANCELLED
-                parent_registration.registered_from.add_log(
-                    action=NodeLog.EMBARGO_CANCELLED,
-                    params={
-                        'node': parent_registration._id,
-                        'embargo_id': parent_registration.embargo._id,
-                    },
-                    auth=Auth(user),
-                )
-                parent_registration.embargo.save()
-            # Ensure retracted registration is public
-            if not parent_registration.is_public:
-                parent_registration.set_privacy('public')
-            parent_registration.update_search()
-            # Retraction status is inherited from the root project, so we
-            # need to recursively update search for every descendant node
-            # so that retracted subrojects/components don't appear in search
-            for node in parent_registration.get_descendants_recursive():
-                node.update_search()
-
-
-def validate_embargo_state(value):
```

Comment:
@HarryRybacki 
I removed the validate_*_state method for Sanctions because as we discussed this is mediated fully internally.